sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,526 @@
+  //
+  // Each node is required to contain a PTransform specification.
+  map<int64, GraphNode> transform_nodes = 1;
+
+  // (Required) A map from pipeline-scoped id to PCollection.
+  map<int64, PCollection> pcollections = 2;
+
+  // (Required) A map from pipeline-scoped id to WindowingStrategy.
+  map<int64, WindowingStrategy> windowing_strategies = 3;
+
+  // (Required) A map from pipeline-scoped id to Coder.
+  map<int64, Coder> coders = 4;
+
+  // (Required) A map from pipeline-scoped id to Environment.
+  map<int64, Environment> environments = 5;
+
+  // (Required) Static display data for the pipeline.
+  DisplayData display_data = 7;
+}
+
+// A generic node in a bipartite directed hierarchical graph.
+message GraphNode {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A URN that describes what kind graph node this is.
+  //
+  // Specifically, the URN should be enough to decipher the inputs,
+  // outputs, and payload.
+  string urn = 2;
+
+  // (Optional) if this node is contained within a composite, a pointer to the
+  // parent.
+  int64 parent_ref = 3;
+
+  // (Required) A map of local input names to target definitions.
+  //
+  // The URN for the graph node may clarify the type of the inputs
+  // (resp. outputs). For example:
+  //
+  //  - in the Runner API these are PCollections
+  //  - in the Fn API they may be Grpc ports
+  //
+  // The payload for this graph node may clarify the relationship of these
+  // inputs. For example:
+  //
+  //  - for a Flatten transform they are merged
+  //  - for a ParDo transform, some may be side inputs
+  //
+  // All inputs are recorded here so that the topological ordering of
+  // the graph is consistent whether or not the payload is understood.
+  //
+  // TODO: fix the value to be int64 as they will always be refs?
+  map<string, google.protobuf.Any> inputs = 4;
+
+  // (Required) A map from local output name to PCollection definitions.
+  //
+  // The URN or payload for the graph node may clarify the type and
+  // relationship of these. For example:
+  //
+  //  - for a ParDo transform, these are tags on PCollections, which will be
+  //    embedded in the DoFn.
+  //  - in the Runner API the targets are PCollections
+  //  - in the Fn API the targets may be Grpc ports
+  //
+  // TODO: fix the value to be int64 as they will always be refs?
+  map<string, google.protobuf.Any> outputs = 5;
+
+  // (Required) A payload determined by the URN, fully specifying this node.
+  google.protobuf.Any payload = 6;
+}
+
+// A PCollection!
+message PCollection {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A unique program-readable name for the value, such as
+  // "foo/baz/bar.out"
+  string label = 2;
+
+  // (Required) The id of the Coder for this PCollection.
+  int64 coder_ref = 3;
+
+  // (Required) Whether this PCollection is bounded or unbounded
+  IsBounded is_bounded = 4;
```

Comment:
This falls into the category of "things that could be inferred, but I'd like the SDK to do so". I think windowing strategy is also in this bucket - there are rules for how it is expected to propagate, but the SDK is the one tasked with implementing those rules.

I guess the main thing a runner does with this information is figure out (a) if it can run the pipeline and (b) how to partition a pipeline into bits that can run batch style versus streaming style. These can probably be done via other means like just looking at Reads and SDFs. Do you have something in mind?

Another reason it is here is that it is a field on Java's `PCollection` object that runners do inspect IIRC. And in a (wildly imagined) mix-and-match scenario a transform might be receiving an input that came from a deserialized pipeline fragment.